examples/pytorch/graphsage/node_classification.py

```diff
@@ -3,130 +3,131 @@
-
-device = 'cuda'
-train_idx = train_idx.to(device)
-valid_idx = valid_idx.to(device)
-test_idx = test_idx.to(device)
-
-graph = graph.to('cuda' if args.pure_gpu else 'cpu')
-
-model = SAGE(graph.ndata['feat'].shape[1], 256, dataset.num_classes).to(device)
-opt = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=5e-4)
-
-sampler = dgl.dataloading.NeighborSampler(
-        [15, 10, 5], prefetch_node_feats=['feat'], prefetch_labels=['label'])
-train_dataloader = dgl.dataloading.DataLoader(
-        graph, train_idx, sampler, device=device, batch_size=1024, shuffle=True,
-        drop_last=False, num_workers=0, use_uva=not args.pure_gpu)
-valid_dataloader = dgl.dataloading.DataLoader(
-        graph, valid_idx, sampler, device=device, batch_size=1024, shuffle=True,
-        drop_last=False, num_workers=0, use_uva=not args.pure_gpu)
-
-durations = []
-for _ in range(10):
-    model.train()
-    t0 = time.time()
-    for it, (input_nodes, output_nodes, blocks) in enumerate(train_dataloader):
-        x = blocks[0].srcdata['feat']
-        y = blocks[-1].dstdata['label']
-        y_hat = model(blocks, x)
-        loss = F.cross_entropy(y_hat, y)
-        opt.zero_grad()
-        loss.backward()
-        opt.step()
-        if it % 20 == 0:
-            acc = MF.accuracy(y_hat, y)
-            mem = torch.cuda.max_memory_allocated() / 1000000
-            print('Loss', loss.item(), 'Acc', acc.item(), 'GPU Mem', mem, 'MB')
-    tt = time.time()
-    print(tt - t0)
-    durations.append(tt - t0)
-
+def batch_evaluate(model, graph, dataloader):
     model.eval()
     ys = []
     y_hats = []
-    for it, (input_nodes, output_nodes, blocks) in enumerate(valid_dataloader):
+    for it, (input_nodes, output_nodes, blocks) in enumerate(dataloader):
         with torch.no_grad():
             x = blocks[0].srcdata['feat']
             ys.append(blocks[-1].dstdata['label'])
             y_hats.append(model(blocks, x))
-    acc = MF.accuracy(torch.cat(y_hats), torch.cat(ys))
-    print('Validation acc:', acc.item())
+    return MF.accuracy(torch.cat(y_hats), torch.cat(ys))
+
+def evaluate(model, graph, device, nid, batch_size):
+    model.eval()
+    nid = nid.to(device)
+    with torch.no_grad():
+        pred = model.inference(graph, device, batch_size)
+        pred = pred[nid].to(device)
+        label = graph.ndata['label'][nid].to(device)
+        return MF.accuracy(pred, label)
+
+def train(args, g, device, data):
+    in_feats, n_classes, train_idx, valid_idx = data
+    train_idx = train_idx.to(device)
+    valid_idx = valid_idx.to(device)
 
-print(np.mean(durations[4:]), np.std(durations[4:]))
+    model = GraphSAGEBatch(in_feats, args.n_hidden, n_classes, args.n_layers, F.relu, args.dropout, args.aggregator_type).to(device)
+    opt = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
+
+    sampler = dgl.dataloading.NeighborSampler([15, 10, 5],
```

Comment:
Fanout should be `[5, 10, 15]`. This has been a mistake for a long time. PyG uses `[15, 10, 5]` because the first fanout is for seed nodes in PyG. https://github.com/pyg-team/pytorch_geometric/blob/6c6e2cc1c350d97eeeab20f698fb1c506aed441e/examples/ogbn_products_sage.py#L21
However, DGL samples fanouts in a reversed order, that is, the last fanout is for seed nodes.

Reply:
Thank you Xin. I will fix and note it, and propagate to other cases.